compiler/backend/src/org/jetbrains/kotlin/codegen/inline/InlineCodegen.kt

```diff
@@ -345,5 +307,26 @@ abstract class InlineCodegen<out T : BaseExpressionCodegen>(
                     InlineUtil.isInlineParameter(field.descriptor) &&
                     InlineUtil.isInline(field.descriptor.containingDeclaration)
         }
+
+        // Stack spilling before inline function call is required if the inlined bytecode has:
+        //   1. try-catch blocks - otherwise the stack spilling before and after them will not be correct;
+        //   2. suspension points - again, the stack spilling around them is otherwise wrong;
+        //   3. loops - OpenJDK cannot JIT-optimize between loop iterations if the stack is not empty.
+        // Instead of checking for loops precisely, we just check if there are any backward jumps -
+        // that is, a jump from instruction #i to instruction #j where j < i.
+        private fun MethodNode.requiresEmptyStackOnEntry(): Boolean = tryCatchBlocks.isNotEmpty() ||
+                instructions.toArray().any { isBeforeSuspendMarker(it) || isBeforeInlineSuspendMarker(it) || isBackwardsJump(it) }
```

Comment:
I'm not familiar with this part. Before there was simple `isSuspend` check above:
```
val shouldSpillStack = isSuspend || !canSkipStackSpillingOnInline(node)
```
Was it excessive? Are new ones  more precise `isBeforeSuspendMarker(it) || isBeforeInlineSuspendMarker(it) ` ? 
cc @ilmirus

Reply:
Yes. The suspend case is similar to the try-catch check - we [spill the stack](https://github.com/JetBrains/kotlin/blob/master/compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt#L460-L462) before suspend calls and restore it afterwards, but the markers that do this are processed before inlining, so the inline function must be entered with an empty stack. It doesn't matter whether the inlined function is suspend if there are no suspend calls in it.`

Reply:
Thank you for explanation!